jvector-base/src/main/java/io/github/jbellis/jvector/pq/ProductQuantization.java

```diff
@@ -214,24 +235,26 @@ private static String arraySummary(float[] a) {
         return "[" + String.join(", ", b) + "]";
     }
 
-    static float[][][] createCodebooks(List<float[]> vectors, int M, int[][] subvectorSizeAndOffset, ForkJoinPool simdExecutor) {
+    static VectorFloat<?>[] createCodebooks(List<VectorFloat<?>> vectors, int M, int[][] subvectorSizeAndOffset, int clusters, ForkJoinPool simdExecutor) {
         return simdExecutor.submit(() -> IntStream.range(0, M).parallel()
                 .mapToObj(m -> {
-                    float[][] subvectors = vectors.stream().parallel()
+                    VectorFloat<?>[] subvectors = vectors.stream().parallel()
                             .map(vector -> getSubVector(vector, m, subvectorSizeAndOffset))
-                            .toArray(float[][]::new);
-                    var clusterer = new KMeansPlusPlusClusterer(subvectors, CLUSTERS, VectorUtil::squareDistance);
+                            .toArray(VectorFloat<?>[]::new);
+                    var clusterer = new KMeansPlusPlusClusterer(subvectors, clusters, VectorUtil::squareDistance);
                     return clusterer.cluster(K_MEANS_ITERATIONS);
                 })
-                .toArray(float[][][]::new))
+                .toArray(VectorFloat<?>[]::new))
                 .join();
     }
     
-    static int closetCentroidIndex(float[] subvector, float[][] codebook) {
+    static int closestCentroidIndex(VectorFloat<?> vector, int[] subvectorSizeAndOffset, VectorFloat<?> codebook) {
         int index = 0;
         float minDist = Integer.MAX_VALUE;
-        for (int i = 0; i < codebook.length; i++) {
-            float dist = VectorUtil.squareDistance(subvector, codebook[i]);
+        // vectorFloat will have n subvectors, each of length subvectorSizeAndOffset[0]
```

Comment:
in the general case it will be subvectorSizeAndOffset.length subvectors, w/ length given in that array

Reply:
basically i was lazy, there are 2N values in that array (abbreviating as "a")

a[n] = size of subvector n
a[n + 1] = offset of subvector n

Reply:
to be clear -- i'm fine throwing away support for unequal subspace sizes, but it was pretty straightforward to support in the Old Regime